core/server/common/src/main/java/alluxio/master/journal/raft/transport/CopycatGrpcConnection.java

```diff
@@ -0,0 +1,439 @@
+   * @param responseObject response object  to send
+   */
+  private void sendResponse(long requestId, ThreadContext context, Object responseObject) {
+    LOG.debug("Sending response of type: {} for request: {}", responseObject.getClass().getName(),
+        requestId);
+    mTargetObserver.onNext(CopycatMessage.newBuilder()
+        .setResponseHeader(CopycatResponseHeader.newBuilder().setRequestId(requestId)
+            .setFailed(responseObject instanceof Throwable))
+        .setMessage(ByteString.copyFrom(context.serializer().writeObject(responseObject).array()))
+        .build());
+  }
+
+  protected void handleResponseMessage(CopycatMessage response) {
+    CopycatGrpcConnection.ContextualFuture future =
+        mResponseFutures.remove(response.getResponseHeader().getRequestId());
+
+    if (future == null) {
+      return;
+    }
+
+    try {
+      if (response.getResponseHeader().getFailed()) {
+        Throwable error = mContext.serializer().readObject(response.getMessage().newInput());
+        LOG.debug("Received error response for request: {}. Error: {}",
+            response.getResponseHeader().getRequestId(), error);
+        future.getContext().executor().execute(() -> future.completeExceptionally(error));
+      } else {
+        Object responseObject = mContext.serializer().readObject(response.getMessage().newInput());
+        LOG.debug("Received response of type: {} request: {}.", responseObject.getClass().getName(),
+            response.getResponseHeader().getRequestId());
+        future.getContext().executor().execute(() -> future
+            .complete(responseObject));
+      }
+    } catch (SerializationException e) {
+      future.getContext().executor().execute(() -> future.completeExceptionally(e));
+    }
+  }
+
+  @Override
+  public Listener<Throwable> onException(Consumer<Throwable> listener) {
+    // Call immediately if the connection was faulted.
+    if (mLastFailure != null) {
+      listener.accept(mLastFailure);
+    }
+    return mExceptionListeners.add(Assert.notNull(listener, "listener"));
+  }
+
+  @Override
+  public Listener<Connection> onClose(Consumer<Connection> listener) {
+    // Call immediately if the connection was closed.
+    if (mClosed) {
+      listener.accept(this);
+    }
+    return mCloseListeners.add(Assert.notNull(listener, "listener"));
+  }
+
+  @Override
+  public CompletableFuture<Void> close() {
+    if (!mClosed) {
+      LOG.debug("Closing connection. Owner: {}", mConnectionOwner);
+      mTargetObserver.onCompleted();
+
+      mClosed = true;
+      // Stop timeout timer.
+      mTimeoutScheduler.cancel();
+
+      // Close outstanding calls.
+      for (CopycatGrpcConnection.ContextualFuture<?> responseFuture : mResponseFutures.values()) {
+        responseFuture.getContext().executor().execute(
+            () -> responseFuture.completeExceptionally(new ConnectException("Connection closed.")));
+      }
+      mResponseFutures.clear();
```

Comment:
What happens if a new response future is added by another thread right before this line is executed?

Reply:
Nice catch. Could cause hangs as copycat trusts transport layer to timeout requests properly.